File: backend/app/schemas/scraper.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List


class ScraperStartRequest(BaseModel):
    campaign_id: str
    account_name: str = Field(default="default", description="The registered session account name")
    search_mode: str = Field(default="COMMENT", description="KEYWORD, HASHTAG, LOCATION, USERNAME, or COMMENT")
    source_query: Optional[str] = Field(default="", description="The query string or post URL")
    post_urls: Optional[List[str]] = Field(default_factory=list, description="List of post URLs for comment scraping")
    keyword_filter: Optional[str] = Field(default=None, description="Keyword filter for comments")
    max_profiles: int = Field(100, description="Max profiles to scrape")
    max_scrolls: int = Field(50, description="Max scrolls to perform")
    include_replies: bool = Field(default=True, description="Whether to include comment replies")
    skip_duplicates: bool = Field(default=True, description="Whether to skip duplicate profiles")
    profile_enrichment: bool = Field(default=True, description="Whether to enrich discovered profiles")
    min_followers: Optional[int] = Field(default=None, description="Minimum followers filter")
    max_followers: Optional[int] = Field(default=None, description="Maximum followers filter")
    min_posts: Optional[int] = Field(default=None, description="Minimum posts filter")
    max_posts: Optional[int] = Field(default=None, description="Maximum posts filter")
    language: Optional[str] = Field(default=None, description="Language filter")
    country: Optional[str] = Field(default=None, description="Country filter")
    business_category: Optional[str] = Field(default=None, description="Business category filter")
    is_business_required: Optional[bool] = Field(default=False, description="Require business account")
    is_verified_required: Optional[bool] = Field(default=False, description="Require verified badge")
    is_email_required: Optional[bool] = Field(default=False, description="Require email found")
    is_phone_required: Optional[bool] = Field(default=False, description="Require phone found")
    is_website_required: Optional[bool] = Field(default=False, description="Require website found")


class ScraperStatusResponse(BaseModel):
    campaign_id: str
    is_running: bool
    status: str = "PENDING"
    stage: str = "Starting"
    current_username: Optional[str] = None
    current_url: Optional[str] = None
    stats: Optional[Dict[str, Any]] = None

File: frontend/src/lib/api/scraper.ts
import { apiClient } from "./client";

export interface ScraperStartPayload {
  campaign_id: string;
  account_name?: string;
  search_mode: string;
  source_query?: string;
  post_urls?: string[];
  keyword_filter?: string | null;
  max_profiles?: number;
  max_scrolls?: number;
  include_replies?: boolean;
  skip_duplicates?: boolean;
  profile_enrichment?: boolean;
  min_followers?: number | null;
  max_followers?: number | null;
  min_posts?: number | null;
  max_posts?: number | null;
  language?: string | null;
  country?: string | null;
  business_category?: string | null;
  is_business_required?: boolean;
  is_verified_required?: boolean;
  is_email_required?: boolean;
  is_phone_required?: boolean;
  is_website_required?: boolean;
}

export interface ScraperJobStats {
  profiles_discovered: number;
  profiles_processed: number;
  profiles_inserted: number;
  duplicates_skipped: number;
  errors: number;
  start_time: number;
  elapsed_time_sec: number;
  status: string;
  stage: string;
  current_username?: string | null;
  current_url?: string | null;
  target_count: number;
}

export interface ScraperStatusResponse {
  campaign_id: string;
  is_running: boolean;
  status: string;
  stage: string;
  current_username?: string | null;
  current_url?: string | null;
  stats?: ScraperJobStats | null;
}

export interface ScraperLogEntry {
  id: string;
  level: "INFO" | "WARNING" | "ERROR" | "CRITICAL";
  message: string;
  created_at: string;
}

export const scraperApi = {
  startScraper: async (payload: ScraperStartPayload) => {
    const { data } = await apiClient.post("/scraper/start", payload);
    return data;
  },

  stopScraper: async (campaignId: string) => {
    const { data } = await apiClient.post(`/scraper/stop/${campaignId}`);
    return data;
  },

  getStatus: async (campaignId: string): Promise<ScraperStatusResponse> => {
    const { data } = await apiClient.get<ScraperStatusResponse>(`/scraper/status/${campaignId}`);
    return data;
  },

  getLogs: async (campaignId: string, limit: number = 100): Promise<ScraperLogEntry[]> => {
    const { data } = await apiClient.get<ScraperLogEntry[]>(`/scraper/logs/${campaignId}`, {
      params: { limit },
    });
    return data;
  },
};
